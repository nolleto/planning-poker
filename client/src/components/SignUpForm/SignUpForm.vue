<template>
  <form @submit.prevent="processForm">
    <div class="field">
      <label class="label">Username</label>
      <input
        type="text"
        class="input"
        name="username"
        v-model="form.username"
      />

      <span class="error">{{ getErrorByName('username') }}</span>
    </div>

    <div class="field">
      <label class="label">Email</label>
      <input
        type="email"
        class="input"
        name="email"
        v-model="form.email"
      />

      <span class="error">{{ getErrorByName('email') }}</span>
    </div>

    <div class="field">
      <label class="label">Password</label>
      <input
        type="password"
        class="input"
        name="password"
        v-model="form.password"
      />

      <span class="error">{{ getErrorByName('password') }}</span>
    </div>

    <div class="field">
      <label class="label">Confirm Password</label>
      <input
        type="password"
        class="input"
        name="passwordConfirmation"
        v-model="form.passwordConfirmation"
      />

      <span class="error">{{ getErrorByName('passwordConfirmation') }}</span>
    </div>

    <div class="field">
      <button type="submit" class="button">Submit</button>
    </div>

    <p v-if="isLoading">Loading...</p>
  </form>
</template>

<script>
export default {
  name: 'SignUpForm',

  props: {
    isLoading: {
      type: Boolean,
      required: true
    },

    getErrorByName: {
      type: Function,
      required: true
    }
  },

  data () {
    return {
      form: {
        username: '',
        email: '',
        password: '',
        passwordConfirmation: ''
      }
    }
  },

  methods: {
    processForm () {
      const data = {
        username: this.form.username,
        email: this.form.email,
        password: this.form.password,
        passwordConfirmation: this.form.passwordConfirmation
      }

      this.$emit('sign-up', data)
    }
  }
}
</script>

<style>

</style>
