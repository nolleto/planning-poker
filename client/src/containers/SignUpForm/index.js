import { connect } from 'vuex-connect'
import SignUpForm from '../../components/SignUpForm'

export default connect({
  gettersToProps: {
    getErrorByName: 'getErrorByName'
  },

  stateToProps: {
    isLoading: ({ isLoading }) => isLoading
  },

  actionsToEvents: {
    'sign-up': (dispatch, data) => dispatch('signUp', data)
  }
})('SignUpFormContainer', SignUpForm)
