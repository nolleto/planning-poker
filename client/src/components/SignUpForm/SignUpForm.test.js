import SignUpForm from './index'
import VeeValidate from 'vee-validate'
import { shallowMount, mount, createLocalVue } from '@vue/test-utils'

const defaultProps = {
  isLoading: false
}

const shallowComponent = (opts = { propsData: defaultProps }) => {
  const localVue = createLocalVue()

  localVue.use(VeeValidate)

  return shallowMount(SignUpForm, {

    ...opts,
    localVue,
    mixins: [{
      methods: {
        fieldIsValid: jest.fn(() => undefined),
        fieldIsInvalid: jest.fn(() => undefined),
        getFieldError: jest.fn(() => undefined),
        getFieldState: jest.fn(() => undefined),
        formHasErrors: jest.fn(() => false)
      }
    }]
  })
}

describe('SignUpForm component', () => {
  let wrapper

  describe('Lifecycle', () => {
    describe('#created', () => {
      const emit = jest.fn()

      beforeAll(() => {
        wrapper = shallowComponent({
          propsData: defaultProps,
          mocks: {
            $emit: emit
          }
        })
        jest.spyOn(wrapper.vm, '$emit')
      })

      it('emits "clean-server-errors"', () => {
        expect(emit).toHaveBeenCalledWith('clean-server-errors')
      })
    })
  })

  describe('Snapshots', () => {
    beforeAll(() => {
      wrapper = shallowComponent()
    })

    it('matches', () => {
      expect(wrapper.element).toMatchSnapshot()
    })

    describe('When is loading', () => {
      beforeAll(() => {
        wrapper = shallowComponent({
          propsData: {
            isLoading: true
          }
        })
      })

      it('matches', () => {
        expect(wrapper.element).toMatchSnapshot()
      })
    })

    describe('When is there is a error', () => {
      beforeAll(() => {
        wrapper = shallowComponent({
          propsData: {
            ...defaultProps,
            serverErrors: {
              email: [ 'invalid' ]
            }
          }
        })
      })

      it('matches', () => {
        expect(wrapper.element).toMatchSnapshot()
      })
    })
  })

  describe('Methods', () => {
    describe('#processForm', () => {
      describe('When form is valid', () => {
        beforeAll(() => {
          wrapper = shallowComponent()
          jest.useFakeTimers()
          jest.spyOn(wrapper.vm, '$emit')
          wrapper.vm.$validator.validateAll = jest.fn(async () => true)
          wrapper.vm.processForm()
        })

        it('calls $validator.validateAll', () => {
          expect(wrapper.vm.$validator.validateAll).toHaveBeenCalled()
        })

        describe('and finished promise', () => {
          beforeAll(() => {
            jest.runAllTimers()
          })

          it('emits "sign-up" with payload', () => {
            expect(wrapper.vm.$emit).toHaveBeenCalledWith(
              'sign-up',
              {
                email: '',
                password: '',
                passwordConfirmation: '',
                username: ''
              }
            )
          })
        })
      })

      describe('When form is invalid', () => {
        beforeAll(() => {
          wrapper = shallowComponent()
          jest.useFakeTimers()
          jest.spyOn(wrapper.vm, '$emit')
          wrapper.vm.$validator.validateAll = jest.fn(async () => false)
          wrapper.vm.processForm()
        })

        it('calls $validator.validateAll', () => {
          expect(wrapper.vm.$validator.validateAll).toHaveBeenCalled()
        })

        describe('and finished promise', () => {
          beforeAll(() => {
            jest.runAllTimers()
          })

          it('does not emits "sign-up" with payload', () => {
            expect(wrapper.vm.$emit).not.toHaveBeenCalled()
          })
        })
      })
    })
  })
})
