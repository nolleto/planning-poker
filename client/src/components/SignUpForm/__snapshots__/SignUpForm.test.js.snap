// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm component Snapshots When is loading matches 1`] = `
<form>
  <div
    class="field"
  >
    <label
      class="label"
    >
      Username
    </label>
     
    <input
      class="input"
      name="username"
      type="text"
    />
     
    <span
      class="error"
    >
      
    </span>
  </div>
   
  <div
    class="field"
  >
    <label
      class="label"
    >
      Email
    </label>
     
    <input
      class="input"
      name="email"
      type="email"
    />
     
    <span
      class="error"
    >
      
    </span>
  </div>
   
  <div
    class="field"
  >
    <label
      class="label"
    >
      Password
    </label>
     
    <input
      class="input"
      name="password"
      type="password"
    />
     
    <span
      class="error"
    >
      
    </span>
  </div>
   
  <div
    class="field"
  >
    <label
      class="label"
    >
      Confirm Password
    </label>
     
    <input
      class="input"
      name="passwordConfirmation"
      type="password"
    />
     
    <span
      class="error"
    >
      
    </span>
  </div>
   
  <div
    class="field"
  >
    <button
      class="button"
      type="submit"
    >
      Submit
    </button>
  </div>
   
  <p>
    Loading...
  </p>
</form>
`;

exports[`SignUpForm component Snapshots matches 1`] = `
<form>
  <div
    class="field"
  >
    <label
      class="label"
    >
      Username
    </label>
     
    <input
      class="input"
      name="username"
      type="text"
    />
     
    <span
      class="error"
    >
      
    </span>
  </div>
   
  <div
    class="field"
  >
    <label
      class="label"
    >
      Email
    </label>
     
    <input
      class="input"
      name="email"
      type="email"
    />
     
    <span
      class="error"
    >
      
    </span>
  </div>
   
  <div
    class="field"
  >
    <label
      class="label"
    >
      Password
    </label>
     
    <input
      class="input"
      name="password"
      type="password"
    />
     
    <span
      class="error"
    >
      
    </span>
  </div>
   
  <div
    class="field"
  >
    <label
      class="label"
    >
      Confirm Password
    </label>
     
    <input
      class="input"
      name="passwordConfirmation"
      type="password"
    />
     
    <span
      class="error"
    >
      
    </span>
  </div>
   
  <div
    class="field"
  >
    <button
      class="button"
      type="submit"
    >
      Submit
    </button>
  </div>
   
  <!---->
</form>
`;
